src/async_impl/response.rs

```diff
@@ -1,23 +1,29 @@
 use std::fmt;
+use std::mem;
+use std::io::{self, Read};
 use std::marker::PhantomData;
 
+use libflate::non_blocking::gzip;
+use tokio_io::AsyncRead;
+use tokio_io::io as async_io;
 use futures::{Async, Future, Poll, Stream};
 use futures::stream::Concat2;
-use header::Headers;
 use hyper::StatusCode;
 use serde::de::DeserializeOwned;
 use serde_json;
 use url::Url;
 
-use super::{body, Body};
+use header::{Headers, ContentEncoding, ContentLength, Encoding, TransferEncoding};
+use super::{decoder, body, Body, Chunk, Decoder};
+use error;
 
 
 /// A Response to a submitted `Request`.
 pub struct Response {
     status: StatusCode,
     headers: Headers,
     url: Url,
-    body: Body,
+    body: Decoder<Body>,
```

Comment:
I think it probably makes sense that the body is a `Decoder` instead, like you've done here. I'd probably remove the generic, and just expose `Decoder` (at least, publicly. If that wraps something that is generic internally, that's fine.)